elastic/kibana
# [Dashboard First] Add to Library Action

## Summary

Counterpart / Opposite of https://github.com/elastic/kibana/pull/74905

Adds an embeddable panel action to add by value embeddables to the library, turning them into by reference embeddables. This action should work with any embeddable that implements the `ReferenceOrValueEmbeddable` interface from #74302.

The action has an implementation in `attribute_service` that comes with a saved object save modal and title duplication detection. 

![Aug-14-2020 18-14-49](https://user-images.githubusercontent.com/14276393/90296657-1b9e3f00-de5a-11ea-84bd-3106f7a00f2d.gif)

## How to test this?

This is just one part of a much larger change, so all changes in this PR are hidden behind a configuration value. In order for the 'by value' workflow to be made default, some architectural changes need to be completed including https://github.com/elastic/kibana/pull/67931, https://github.com/elastic/kibana/issues/71499, and https://github.com/elastic/kibana/issues/61663

1. you'll need to set the `allowByValueEmbeddables` config value to `true`:
https://github.com/elastic/kibana/blob/5b64a4cdfea92639ddf9e23c86c43dda5ea3ac68/src/plugins/dashboard/config.ts#L23

2. There are no production ready embeddables that implement the `ReferenceOrValueEmbeddable` yet so you will have to start the developer examples. `yarn start --run-examples`

3. Open a dashboard, and use the add panels functionality to add a Book embeddable. You should be able to then unlink it from its library item. using the 'unlink book from library item' action menu item.

4. Use the 'Add to Library' action, being sure to save the embeddable with a unique name.

### Checklist


### For maintainers
